### ISSUE TITLE
[Rust] Support extraEnv under cargo settings

### ISSUE THREAD

**Issue by setzer22** [emacs-lsp/lsp-mode#3963]
Rust-analyzer documents an `extraEnv` setting under the `cargo` section (https://rust-analyzer.github.io/manual.html#toolchain):

> rust-analyzer.cargo.extraEnv (default: {})
>
>    Extra environment variables that will be set when running cargo, rustc or other commands within the workspace. Useful for setting RUSTFLAGS.

But I don't see a way to access it in lsp-mode. A quick check in `lsp-rust.el` shows there's no variable to allow setting this property:
```lisp
    :cargo (:allFeatures ,(lsp-json-bool lsp-rust-all-features)
            :noDefaultFeatures ,(lsp-json-bool lsp-rust-no-default-features)
            :features ,lsp-rust-features
            :target ,lsp-rust-analyzer-cargo-target
            :runBuildScripts ,(lsp-json-bool lsp-rust-analyzer-cargo-run-build-scripts)
            ; Obsolete, but used by old Rust-Analyzer versions
            :loadOutDirsFromCheck ,(lsp-json-bool lsp-rust-analyzer-cargo-run-build-scripts)
            :autoreload ,(lsp-json-bool lsp-rust-analyzer-cargo-auto-reload)
            :useRustcWrapperForBuildScripts ,(lsp-json-bool lsp-rust-analyzer-use-rustc-wrapper-for-build-scripts)
            :unsetTest ,lsp-rust-analyzer-cargo-unset-test)
```

This would be an easy fix, just a matter of adding a new `lsp-rust-analyzer-*` customizable variable and include it there.

I would submit the PR myself, but there's something that's not entirely clear to me. The JSON expects a dictionary for this field, mapping env var names to their values. What should the type of the variable on the emacs side be? A list of cons cells? e.g. `'(("RUSTFLAGS" . "something something) ("CARGO_TARGET_DIR". "/abc/def") ...)`. Also, I see there's `lsp-json-bool`, but is there an alternative function to convert something to a JSON dictionary?

If you can provide some guidance, I'd be happy to contribute a fix :smile:

**Comment by yyoncho** [comment]
> The JSON expects a dictionary for this field, mapping env var names to their values. What should the type of the variable on the emacs side be?

alist, plist, hash-table will do the job. Typically plist or alist.

**Comment by setzer22** [comment]
Thanks @yyoncho. Do you mean these things should already encode as JSON dictionaries right away?

**Comment by yyoncho** [comment]
I am not sure I understand 100% the question. You should use elisp structures that will be serialized to json map. These are the aforementioned.

**Comment by setzer22** [comment]
Thanks, that's exactly what I was asking, whether the elisp data structures would serialize to JSON :smile:  Sorry if I wasn't clear enough

**Comment by jbr** [comment]
Is this resolved? Is a PR still needed? I'd love to supply extraEnv, ideally on a per-workspace basis

**Comment by jcs090218** [comment]
Resolved in #4015.
